xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -222,79 +234,247 @@ private static LdsUpdate processClientSideListener(Listener listener, boolean pa
       throw new ResourceInvalidException(
           "Could not parse HttpConnectionManager config from ApiListener", e);
     }
+    return LdsUpdate.forApiListener(
+        parseHttpConnectionManager(hcm, parseFilter, true /* forClient */));
+  }
+
+  private LdsUpdate processServerSideListener(Listener proto, boolean parseFilter)
+      throws ResourceInvalidException {
+    return LdsUpdate.forTcpListener(
+        parseServerSideListener(proto, tlsContextManager, parseFilter));
+  }
+
+  @VisibleForTesting
+  static EnvoyServerProtoData.Listener parseServerSideListener(
+      Listener proto, TlsContextManager tlsContextManager, boolean parseFilter)
+      throws ResourceInvalidException {
+    if (!proto.getTrafficDirection().equals(TrafficDirection.INBOUND)) {
+      throw new ResourceInvalidException(
+          "Listener " + proto.getName() + " with invalid traffic direction: "
+              + proto.getTrafficDirection());
+    }
+    if (!proto.getListenerFiltersList().isEmpty()) {
+      throw new ResourceInvalidException(
+          "Listener " + proto.getName() + " cannot have listener_filters");
+    }
+    if (proto.hasUseOriginalDst()) {
+      throw new ResourceInvalidException(
+          "Listener " + proto.getName() + " cannot have use_original_dst set to true");
+    }
+
+    String address = null;
+    if (proto.hasAddress() && proto.getAddress().hasSocketAddress()) {
```

Comment:
- `listener.address` is a required field as per the `validate.rules` annotation on the field
- isn't it sufficient to just check for `proto.getAddress().hasSocketAddress()`?

Reply:
> `listener.address` is a required field as per the `validate.rules` annotation on the field

I've previously asked Mark regarding the necessity of checking _required_ fields, he was arguing that we shouldn't assume the proto validate plugin is used by the control plane (while I do have some disagreement for letting gRPC implementation take this _unnecessary_ burden) and it's still better to always check by ourselves.

> isn't it sufficient to just check for `proto.getAddress().hasSocketAddress()`?

That sounds sufficient. I can change it.

Reply:
Updated.